library/core/class.oauth2.php

```diff
@@ -0,0 +1,792 @@
+
+        // Throw an event so that other plugins can add/remove stuff from the basic sso.
+        $sender->fireEvent('OAuth');
+
+        SpamModel::disabled(true);
+        $sender->setData('Trusted', true);
+        $sender->setData('Verified', true);
+    }
+
+
+    /**
+     * Request access token from provider.
+     *
+     * @param string $code code returned from initial handshake with provider.
+     *
+     * @return mixed Result of the API call to the provider, usually JSON.
+     */
+    public function requestAccessToken($code) {
+        $provider = $this->provider();
+        $uri = $provider['TokenUrl'];
+
+        $defaultParams = [
+            'code' => $code,
+            'client_id' => $provider['AssociationKey'],
+            'redirect_uri' => url('/entry/'. $this->getProviderKey(), true),
+            'client_secret' => $provider['AssociationSecret'],
+            'grant_type' => 'authorization_code',
+            'scope' => $this->scope
+        ];
+
+        $post = array_merge($defaultParams, $this->requestAccessTokenParams);
+
+        $this->log('Before calling API to request access token', ['requestAccessToken' => ['targetURI' => $uri, 'post' => $post]]);
+
+        return $this->api($uri, 'POST', $post);
+    }
+
+
+    /**
+     *   Allow the admin to input the keys that their service uses to send data.
+     *
+     * @param array $rawProfile profile as it is returned from the provider.
+     *
+     * @return array Profile array transformed by child class or as is.
+     */
+    public function translateProfileResults($rawProfile = []) {
+        $provider = $this->provider();
+        $translatedKeys = [
+            val('ProfileKeyEmail', $provider, 'email') => 'Email',
+            val('ProfileKeyPhoto', $provider, 'picture') => 'Photo',
+            val('ProfileKeyName', $provider, 'displayname') => 'Name',
+            val('ProfileKeyFullName', $provider, 'name') => 'FullName',
+            val('ProfileKeyUniqueID', $provider, 'user_id') => 'UniqueID'
+        ];
+
+        $profile = arrayTranslate($rawProfile, $translatedKeys, true);
+
+        $profile['Provider'] = $this->providerKey;
+
+        return $profile;
+    }
+
+
+    /**
+     * Get profile data from authentication provider through API.
+     *
+     * @return array User profile from provider.
+     */
+    public function getProfile() {
+        $provider = $this->provider();
+
+        $uri = $this->requireVal('ProfileUrl', $provider, 'provider');
+
+        $defaultParams = array(
+            'access_token' => $this->accessToken()
+        );
+
+        $requestParams = array_merge($defaultParams, $this->profileRequestParams);
+
+        $rawProfile = $this->api($uri, 'GET', $requestParams);
+
+        $profile = $this->translateProfileResults($rawProfile);
+
+        $this->log('getProfile API call', ['ProfileUrl' => $uri, 'Params' => $requestParams, 'RawProfile' => $rawProfile, 'Profile' => $profile]);
+
+        return $profile;
```

Comment:
Could probably make this method less breezy (linebreaks). Getting a little carried away now.